### PR TITLE
Feature/storage markets types

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Add markets protocols and interfaces.

### CODE REVIEW

File: core/markets/storage/node_api/storage_provider_node.hpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  using primitives::DealId;
+  using primitives::PieceDescriptor;
+  using primitives::TokenAmount;
+  using primitives::address::Address;
+
+  using DealSectorCommittedCallback =
+      std::function<void(outcome::result<void>)>;
+
+  /// Part of Fuhon node interface to be called by StorageProvider
+  class StorageProviderNode {
```
Comment (igor-egorov) [inline core/markets/storage/node_api/storage_provider_node.hpp:37]: Not needed and could be removed.

Comment (Alexey-N-Chernyshov) [reply core/markets/storage/node_api/storage_provider_node.hpp]: removed

File: core/markets/storage/stored_ask.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  outcome::result<void> StoredAsk::addAsk(const TokenAmount &price,
+                                          ChainEpoch duration) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    uint64_t sequential_number{0};
+    if (signed_storage_ask_) {
+      sequential_number = signed_storage_ask_->ask.seq_no + 1;
+    }
+
+    OUTCOME_TRY(chain_head, api_->ChainHead());
+    auto &&height{chain_head.height};
```
Comment (turuslan) [inline core/markets/storage/stored_ask.cpp:19]: Remove `auto &&height{chain_head.height};`, or change `ask.timestamp = chain_head.height;`

File: core/markets/storage/stored_ask.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace fc::markets::storage {
+
+  outcome::result<void> StoredAsk::addAsk(const TokenAmount &price,
+                                          ChainEpoch duration) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    uint64_t sequential_number{0};
+    if (signed_storage_ask_) {
+      sequential_number = signed_storage_ask_->ask.seq_no + 1;
+    }
```
Comment (turuslan) [inline core/markets/storage/stored_ask.cpp:16]: Move to `ask.seq_no = signed_storage_ask_ ? signed_storage_ask_->ask.seq_no + 1 : 0;`

File: core/markets/storage/network/libp2p_storage_market_network.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        peer_info,
+        kAskProtocolId,
+        [&stream](outcome::result<std::shared_ptr<Stream>> new_stream) {
+          stream = std::move(new_stream.value());
+        });
+    return std::make_shared<Libp2pAskStream>(peer_id, stream);
+  }
+
+  outcome::result<std::shared_ptr<StorageDealStream>>
+  Libp2pStorageMarketNetwork::newDealStream(const PeerId &peer_id) {
```
Comment (turuslan) [inline core/markets/storage/network/libp2p_storage_market_network.cpp:31]: Connection creation is asynchronous and should use callback.
`outcome::result` can't return asynchronous errors and can't notify when connection is established.

File: core/markets/storage/ask_stream.hpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#include "common/outcome.hpp"
+#include "markets/storage/ask_protocol.hpp"
+
+namespace fc::markets::storage {
+
+  /**
+   * StorageAskStream is a stream for reading/writing requests & responses on
+   * the Storage Ask protocol
+   */
+  class StorageAskStream {
```
Comment (turuslan) [inline core/markets/storage/ask_stream.hpp:18]: Remove this and use `CborStream` from #166.
It provides `read<T>`/`write<T>` to read/write cbor objects.
Also it exposes underlying `Stream`, so `close` can be called directly.

File: core/markets/storage/deal_protocol.hpp [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+             client,
+             state,
+             piece_path,
+             metadata_path,
+             connection_closed,
+             message,
+             ref,
+             deal_id)
+
+  struct ClientDeal : public ClientDealProposal {
```
Comment (turuslan) [inline core/markets/storage/deal_protocol.hpp:112]: Use composition, not inheritance.
Also it should be encoded/deccoded as first element.

File: core/markets/storage/deal_protocol.hpp [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+             miner_worker,
+             deal_id,
+             data_ref,
+             message,
+             publish_message)
+
+  /**
+   * StorageDeal is a local combination of a proposal and a current deal state
+   */
+  struct StorageDeal : public DealProposal, public DealState {};
```
Comment (turuslan) [inline core/markets/storage/deal_protocol.hpp:136]: If encoding/decoding required, use composition instead of inheritance.

Comment (Alexey-N-Chernyshov) [reply core/markets/storage/deal_protocol.hpp]: done

File: core/markets/storage/deal_protocol.hpp (hunk 6 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/deal_protocol.hpp]: done

File: core/markets/storage/ask_stream.hpp (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/ask_stream.hpp:18]: 👍

File: core/markets/storage/network/libp2p_storage_market_network.cpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply core/markets/storage/network/libp2p_storage_market_network.cpp]: added callback handler

**Pull request merged**